src/cargo/ops/cargo_compile.rs

```diff
@@ -1625,7 +1612,13 @@ fn traverse_and_share(
         .map(|dep| {
             let new_dep_unit =
                 traverse_and_share(interner, memo, new_graph, unit_graph, &dep.unit, to_host);
-            new_dep_unit.hash(&mut dep_hash);
+
+            // The default hash for a unit is the interned pointer, but we need dep_hash to
+            // be stable across Cargo sessions for Doscrape units to be memoized, so we
+            // take care to hash the contents of the unit here.
+            let inner: &UnitInner = &*new_dep_unit;
+            inner.hash(&mut dep_hash);
```

Comment:
Ok yes, I think metadata is actually what I needed. So I changed this feature to use `metadata` instead of `buildkey`, and I also added a test to ensure that scraping is properly memoized. I reverted all changed to unit hashing, since they're no longer necessary.